README.md
# videojs-shuttle-controls

Adds shuttle controls(JKL controls) to video.js
[Live Demo](https://sweetberry.github.io/videojs-shuttle-controls/) 

## Table of Contents

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [Installation](#installation)
- [Usage](#usage)
  - [`<script>` Tag](#script-tag)
  - [Browserify/CommonJS](#browserifycommonjs)
  - [RequireJS/AMD](#requirejsamd)
  - [ES6](#es6)
- [Options](#options)
- [License](#license)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
## Installation

```sh
npm install --save videojs-shuttle-controls
```

## Usage

To include videojs-shuttle-controls on your website or web application, use any of the following methods.

### `<script>` Tag

This is the simplest case. Get the script in whatever way you prefer and include the plugin _after_ you include [video.js][videojs], so that the `videojs` global is available.

```html
<script src="//path/to/video.min.js"></script>
<script src="//path/to/videojs-shuttle-controls.min.js"></script>
<script>
  var player = videojs('my-video', {
    playbackRates: [-10, -5, -2, -1, -0.5, 0.5, 1, 2, 5, 10]
  });

  player.shuttleControls({
    playPauseKeys       : [' ', 'k'],
    backwardKeys        : ['j'],
    forwardKey          : ['l'],
    inKey               : ['i'],
    outKey              : ['o'],
    backwardFrameKey    : ['ArrowLeft'],
    forwardFrameKey     : ['ArrowRight'],
    shiftMagnification  : 10,
    fps                 : 30
  });
</script>
```

### Browserify/CommonJS

When using with Browserify, install videojs-shuttle-controls via npm and `require` the plugin as you would any other module.

```js
var videojs = require('video.js');

// The actual plugin function is exported by this module, but it is also
// attached to the `Player.prototype`; so, there is no need to assign it
// to a variable.
require('videojs-shuttle-controls');

var player = videojs('my-video', {
  playbackRates: [-10, -5, -2, -1, -0.5, 0.5, 1, 2, 5, 10]
});

player.shuttleControls({
  playPauseKeys     : [' ', 'k'],
  backwardKeys      : ['j'],
  forwardKey        : ['l'],
  inKey             : ['i'],
  outKey            : ['o'],
  backwardFrameKey  : ['ArrowLeft'],
  forwardFrameKey   : ['ArrowRight'],
  shiftMagnification: 10,
  fps               : 30
});
```

### RequireJS/AMD

When using with RequireJS (or another AMD library), get the script in whatever way you prefer and `require` the plugin as you normally would:

```js
require(['video.js', 'videojs-shuttle-controls'], function(videojs) {
  var player = videojs('my-video', {
    playbackRates: [-10, -5, -2, -1, -0.5, 0.5, 1, 2, 5, 10]
  });

  player.shuttleControls({
    playPauseKeys       : [' ', 'k'],
    backwardKeys        : ['j'],
    forwardKey          : ['l'],
    inKey               : ['i'],
    outKey              : ['o'],
    backwardFrameKey    : ['ArrowLeft'],
    forwardFrameKey     : ['ArrowRight'],
    shiftMagnification  : 10,
    fps                 : 30
  });
  
});
```

### ES6
```js
import videojs from 'video.js'
import 'videojs-shuttle-controls'

const player = videojs('my-video', {
  playbackRates: [-10, -5, -2, -1, -0.5, 0.5, 1, 2, 5, 10]
});

player.shuttleControls({
  playPauseKeys     : [' ', 'k'],
  backwardKeys      : ['j'],
  forwardKey        : ['l'],
  inKey             : ['i'],
  outKey            : ['o'],
  backwardFrameKey  : ['ArrowLeft'],
  forwardFrameKey   : ['ArrowRight'],
  shiftMagnification: 10,
  fps               : 30
});
```


## Options

- `playPauseKeys` (array of string): The keys to toggle between play and pause (default: `[' ', 'k']`)
- `backwardKeys` (array of string): The keys to play backward (default: `['j']`)
- `forwardKey` (array of string): The keys to play forward (default: `['l']`)
- `inKey` (array of string): The keys to confirm the in point of loop
- `outKey` (array of string): The keys to confirm the out point of loop
- `backwardFrameKey` (array of string): The keys to step backward (default: `['ArrowLeft']`)
- `forwardFrameKey` (array of string): The keys to step forward (default: `['ArrowRight']`)
- `shiftMagnification` (number): The step magnification when using shift key (default: `10`)
- `fps` (number): The video frame rate (default: `30`)
- `playbackRates` (array of number): Set playbackRates. If you want to display in the control bar, set it to the player's playbackRates instead of here. (default: `[-10, -5, -2, -1, -0.5, 0.5, 1, 2, 5, 10]`) 




## License

MIT. Copyright (c) sweetberry &lt;pixel@sweetberry.com&gt;


[videojs]: http://videojs.com/




CHANGELOG.md
<a name="1.2.5"></a>
## [1.2.5](https://github.com/sweetberry/videojs-shuttle-controls/compare/v1.2.4...v1.2.5) (2023-06-15)

<a name="1.2.4"></a>
## [1.2.4](https://github.com/sweetberry/videojs-shuttle-controls/compare/v1.2.3...v1.2.4) (2020-04-26)

<a name="1.2.3"></a>
## [1.2.3](https://github.com/sweetberry/videojs-shuttle-controls/compare/v1.2.2...v1.2.3) (2020-04-26)

<a name="1.2.2"></a>
## [1.2.2](https://github.com/sweetberry/videojs-shuttle-controls/compare/v1.2.1...v1.2.2) (2020-04-26)

<a name="1.2.1"></a>
## [1.2.1](https://github.com/sweetberry/videojs-shuttle-controls/compare/v1.1.1...v1.2.1) (2020-04-26)

<a name="1.2.0"></a>
## [1.2.0](https://github.com/sweetberry/videojs-shuttle-controls/compare/v1.1.1...v1.2.0) (2020-04-26)
- Add in-out loop
<a name="1.1.1"></a>
## 1.1.1 (2019-09-18)

<a name="1.1.0"></a>
## 1.1.0 (2019-08-25)
- Add frame by frame controls
<a name="1.0.3"></a>
## 1.0.3 (2019-08-21)
-　Fix deprecated this.options ()
<a name="1.0.1"></a>
## 1.0.1 (2019-08-20)
- Change README



test/plugin.test.js
import document from 'global/document';

import QUnit from 'qunit';
import sinon from 'sinon';
import videojs from 'video.js';

import plugin from '../src/plugin';

const Player = videojs.getComponent('Player');

QUnit.test('the environment is sane', function(assert) {
  assert.strictEqual(typeof Array.isArray, 'function', 'es5 exists');
  assert.strictEqual(typeof sinon, 'object', 'sinon exists');
  assert.strictEqual(typeof videojs, 'function', 'videojs exists');
  assert.strictEqual(typeof plugin, 'function', 'plugin is a function');
});

QUnit.module('videojs-shuttle-controls', {

  beforeEach() {

    // Mock the environment's timers because certain things - particularly
    // player readiness - are asynchronous in video.js 5. This MUST come
    // before any player is created; otherwise, timers could get created
    // with the actual timer methods!
    this.clock = sinon.useFakeTimers();

    this.fixture = document.getElementById('qunit-fixture');
    this.video = document.createElement('video');
    this.fixture.appendChild(this.video);
    this.player = videojs(this.video);
  },

  afterEach() {
    this.player.dispose();
    this.clock.restore();
  }
});

QUnit.test('registers itself with video.js', function(assert) {
  assert.expect(2);

  assert.strictEqual(
    typeof Player.prototype.shuttleControls,
    'function',
    'videojs-shuttle-controls plugin was registered'
  );

  this.player.shuttleControls();

  // Tick the clock forward enough to trigger the player to be "ready".
  this.clock.tick(1);

  assert.ok(
    this.player.hasClass('vjs-shuttle-controls'),
    'the plugin adds a class to the player'
  );
});


src/plugin.js
import videojs from 'video.js';
import {version as VERSION} from '../package.json';

const Plugin = videojs.getPlugin('plugin');

// Default options for the plugin.
const defaults = {
  playPauseKeys: [' ', 'k'],
  backwardKeys: ['j'],
  forwardKey: ['l'],
  backwardFrameKey: ['ArrowLeft'],
  forwardFrameKey: ['ArrowRight'],
  inKey: ['i'],
  outKey: ['o'],
  inTime: null,
  outTime: null,
  shiftMagnification: 10,
  fps: 30,
  playbackRates: [-10, -5, -2, -1, -0.5, 0.5, 1, 2, 5, 10]
};

/**
 * @param {Promise} value
 */
function silencePromise(value) {
  if (value !== null && typeof value.then === 'function') {
    value.then(null, function(e) {
    });
  }
}

/**
 * An advanced Video.js plugin. For more information on the API
 *
 * See: https://blog.videojs.com/feature-spotlight-advanced-plugins/
 */
class ShuttleControls extends Plugin {

  /**
   * Create a ShuttleControls plugin instance.
   *
   * @param  {Player} player
   *         A Video.js Player instance.
   *
   * @param  {Object} [options] An optional options object.
   * @param  {String[]} [options.playPauseKeys]
   * @param  {String[]} [options.backwardKeys]
   * @param  {String[]} [options.forwardKey]
   * @param  {String[]} [options.backwardFrameKey]
   * @param  {String[]} [options.forwardFrameKey]
   * @param  {String[]} [options.inKey]
   * @param  {String[]} [options.outKey]
   * @param  {!number} [options.inTime]
   * @param  {!number} [options.outTime]
   * @param  {number} [options.shiftMagnification]
   * @param  {number} [options.fps]
   * @param  {Number[]} [options.playbackRates]
   */
  constructor(player, options) {
    // the parent class will add player under this.player
    super(player);

    this.options = videojs.mergeOptions(defaults, options);

    this.player.ready(() => {
      this.player.addClass('vjs-shuttle-controls');
    });

    this.playPauseKeys = this.options.playPauseKeys;
    this.backwardKeys = this.options.backwardKeys;
    this.forwardKey = this.options.forwardKey;
    this.backwardFrameKey = this.options.backwardFrameKey;
    this.forwardFrameKey = this.options.forwardFrameKey;
    this.inKey = this.options.inKey;
    this.outKey = this.options.outKey;
    this.inTime = this.options.inTime;
    this.outTime = this.options.outTime;
    this.shiftMagnification = this.options.shiftMagnification;

    this.fps = this.options.fps;

    this.currentPlaybackRate = 1;
    this.isPlaying = false;

    this.player.on('keydown', (e) => {
      this._onKeydown(e);
    });
    this.playbackRates = this.options.playbackRates;
    if (player.options_.playbackRates.length) {
      this.playbackRates = player.options_.playbackRates;
    }

    this.player.on('ratechange', () => {
      const speed = this.player.playbackRate();

      if (this.isPlaying) {
        this._play(speed);
      } else {
        this.player.$('.vjs-playback-rate-value').innerHTML = `${speed}x`;
        this.currentPlaybackRate = speed;
      }

    });
    this.player.one('play', () => {
      this.isPlaying = true;
    });
    this._tweakMenuStyles();
    this._listenPlayControlButton();
    this._createMarkers();
  }

  _tweakMenuStyles() {
    const liArray = videojs.dom.$$('.vjs-menu li');

    liArray.forEach((li) => {
      li.style['font-size'] = 'inherit';
      li.style.padding = '0';
    });
  }

  _listenPlayControlButton() {
    const buttonDom = videojs.dom.$('.vjs-play-control');
    const techDom = videojs.dom.$('.vjs-tech');
    const playControlHandler = () => {
      if (techDom.parentNode.id == this.player.id) {
        const hasPaused = videojs.dom.hasClass(techDom.parentNode, 'vjs-paused');
        const currentPlaybackRate = this.currentPlaybackRate;
        const isNegativePlaying = this.isPlaying && currentPlaybackRate < 0 && hasPaused;
        const isNegativePause = !this.isPlaying && currentPlaybackRate < 0 && hasPaused;
        const isPositivePlaying = this.isPlaying && currentPlaybackRate > 0 && !hasPaused;
        const isPositivePause = !this.isPlaying && currentPlaybackRate > 0 && hasPaused;

        if (isNegativePlaying || isPositivePlaying) {
          this._playPause();
        }
        if (isNegativePause || isPositivePause) {
          this._play(currentPlaybackRate);
        }
      }
    };

    buttonDom.addEventListener('click', playControlHandler);
    techDom.addEventListener('click', playControlHandler);
  }

  _onKeydown(e) {
    this._clearNegativeTimer();
    this._clearLoopTimer();

    const mag = e.shiftKey ? this.shiftMagnification : 1;

    if (this.playPauseKeys.indexOf(e.key) > -1) {
      this._togglePlay();
    }
    if (this.forwardKey.indexOf(e.key) > -1) {
      this._shuttlePlaybackRate(1);
    }
    if (this.backwardKeys.indexOf(e.key) > -1) {
      this._shuttlePlaybackRate(-1);
    }
    if (this.forwardFrameKey.indexOf(e.key) > -1) {
      this.frameByFrame(mag);
    }
    if (this.backwardFrameKey.indexOf(e.key) > -1) {
      this.frameByFrame(-mag);
    }
    if (this.inKey.indexOf(e.key) > -1) {
      this._updateInTime();
      this._updateMarkers();
    }
    if (this.outKey.indexOf(e.key) > -1) {
      this._updateOutTime();
      this._updateMarkers();
    }
  }

  _updateInTime() {
    const isPlaying = this.isPlaying;
    const currentPlaybackRate = this.currentPlaybackRate;

    this.inTime = this.player.currentTime();
    if (this.outTime && this.inTime > this.outTime) {
      this.outTime = this.inTime;
    }

    if (isPlaying && currentPlaybackRate < 0) {
      this.player.currentTime(this.outTime || this.player.duration());
    }

    if (isPlaying) {
      this._play(currentPlaybackRate);
    }
  }

  _updateOutTime() {
    const isPlaying = this.isPlaying;
    const currentPlaybackRate = this.currentPlaybackRate;

    this.outTime = this.player.currentTime();

    if (this.inTime && this.inTime > this.outTime) {
      this.inTime = this.outTime;
    }

    if (isPlaying && currentPlaybackRate > 0) {
      this.player.currentTime(this.inTime || 0);
    }

    if (isPlaying) {
      this._play(currentPlaybackRate);
    }
  }

  _playPause() {
    const player = this.player;

    this._clearLoopTimer();
    this._clearNegativeTimer();
    this.isPlaying = false;
    // console.log(this.isPlaying);
    const buttonDom = videojs.dom.$('.vjs-play-control');

    if (videojs.dom.hasClass(buttonDom, 'vjs-playing')) {
      videojs.dom.toggleClass(buttonDom, 'vjs-paused');
      videojs.dom.toggleClass(buttonDom, 'vjs-playing');
    }
    player.pause();
  }

  _ResetPlaybackRate() {
    const player = this.player;

    this.currentPlaybackRate = 1;
    player.playbackRate(1);
    player.$('.vjs-playback-rate-value').innerHTML = '1x';
  }

  _togglePlay() {
    const player = this.player;

    if (!this.isPlaying) {
      this._play(1);
      silencePromise(player.play());
    } else {
      this._playPause();
      this._ResetPlaybackRate();
    }
  }

  _negativePlay(speed) {
    const player = this.player;
    const frameTime = 1 / (this.fps);
    const buttonDom = videojs.dom.$('.vjs-play-control');

    if (videojs.dom.hasClass(buttonDom, 'vjs-paused')) {
      videojs.dom.toggleClass(buttonDom, 'vjs-paused');
      videojs.dom.toggleClass(buttonDom, 'vjs-playing');
    }
    this.isPlaying = true;
    player.pause();
    player.$('.vjs-playback-rate-value').innerHTML = `${speed}x`;
    let newTime = this.player.currentTime() + frameTime * speed;

    if (newTime < 0) {
      newTime = player.duration();
    }
    player.currentTime(newTime);

  }

  _loop(speed) {
    const player = this.player;
    const inTime = this.inTime || 0;
    const outTime = this.outTime || player.duration();
    const currentTime = this.player.currentTime();
    const frameTime = 1 / (this.fps);

    if (this.isPlaying && inTime === outTime) {
      this._playPause();
      this._ResetPlaybackRate();
      player.currentTime(inTime);
    } else if (this.isPlaying && speed > 0) {
      if (currentTime < inTime) {
        player.currentTime(inTime + frameTime * speed);
      } else if (currentTime > outTime) {
        player.currentTime(inTime + frameTime * speed);
      }
    } else if (this.isPlaying) {
      if (currentTime < inTime) {
        player.currentTime(outTime + frameTime * speed);
      } else if (currentTime > outTime) {
        player.currentTime(outTime + frameTime * speed);
      }
    }

  }

  _play(speed) {
    this.isPlaying = true;

    const player = this.player;
    const frameTime = 1 / (this.fps);

    this.currentPlaybackRate = speed;
    this._clearLoopTimer();
    this._clearNegativeTimer();
    if (speed > 0) {
      player.playbackRate(speed);
      silencePromise(player.play());
    } else {
      this.negativeTimer = setInterval(() => {
        this._negativePlay(speed);
      }, frameTime * 1000);
    }

    this.loopTimer = setInterval(() => {
      this._loop(speed);
    }, frameTime * 1000);
  }

  _clearLoopTimer() {
    if (this.loopTimer) {
      clearInterval(this.loopTimer);
    }
  }

  _clearNegativeTimer() {
    if (this.negativeTimer) {
      clearInterval(this.negativeTimer);
    }
  }

  _shuttlePlaybackRate(step) {
    if (!this.isPlaying) {
      this._play(step);
    } else {
      const playbackRates = this.playbackRates;
      const nowRate = this.currentPlaybackRate;
      const nowIndex = playbackRates.indexOf(nowRate);
      const nextIndex = Math.min(Math.max(nowIndex + step, 0), playbackRates.length - 1);

      this._play(playbackRates[nextIndex]);
    }
  }

  frameByFrame(step) {
    const player = this.player;
    const frameTime = 1 / this.fps;

    player.pause();
    player.currentTime(this.player.currentTime() + frameTime * step);
  }

  _createMarkers() {
    const player = this.player;
    const inMarkerStyle = {
      'position': 'absolute',
      'width': '0',
      'height': '0',
      'display': 'inline-block',
      'border-style': 'solid',
      'border-width': ' 0 5px 10px 0',
      'border-color': 'transparent #ffffff transparent transparent',
      'bottom': '5px',
      'margin-left': '-5px',
      'left': '0%'
    };
    const outMarkerStyle = {
      'position': 'absolute',
      'width': '0',
      'height': '0',
      'display': 'inline-block',
      'border-style': 'solid',
      'border-width': '10px 5px 0 0',
      'border-color': '#ffffff transparent transparent transparent',
      'bottom': '5px',
      'left': '100%'
    };
    const inMarker = videojs.dom.createEl('div', {}, {
      id: 'vjs_shuttleControls_inMarker'
    });

    const outMarker = videojs.dom.createEl('div', {}, {
      id: 'vjs_shuttleControls_outMarker'
    });

    Object.keys(inMarkerStyle).forEach((key) => {
      inMarker.style[key] = inMarkerStyle[key];
    });

    Object.keys(outMarkerStyle).forEach((key) => {
      outMarker.style[key] = outMarkerStyle[key];
    });

    player.el().querySelector('.vjs-progress-holder')
      .appendChild(inMarker);
    player.el().querySelector('.vjs-progress-holder')
      .appendChild(outMarker);

    this.inMarker = inMarker;
    this.outMarker = outMarker;
  }

  _updateMarkers() {
    const player = this.player;
    const inTime = this.inTime || 0;
    const outTime = this.outTime || player.duration();
    const inPosition = (inTime / player.duration()) * 100;
    const outPosition = (outTime / player.duration()) * 100;
    const inMarker = this.inMarker;
    const outMarker = this.outMarker;

    inMarker.style.left = `${inPosition}%`;
    outMarker.style.left = `${outPosition}%`;
  }

}

// Define default values for the plugin's `state` object here.
ShuttleControls.defaultState = {};

// Include the version number.
ShuttleControls.VERSION = VERSION;

// Register the plugin with video.js.
videojs.registerPlugin('shuttleControls', ShuttleControls);

export default ShuttleControls;
